test/unit/services/aws/credentials_service_test.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+class Aws::CredentialsServiceTest < ActiveSupport::TestCase
+  test '#call returns IAM credentials when available' do
+    assume_role_web_identity_service_instance.expects(:identity_credentials).never
+
+    assert Aws::CredentialsService.call(iam_auth_params), iam_auth_params
+  end
+
+  test '#call returns STS credentials when available' do
+    assume_role_web_identity_service_instance
+      .expects(:config)
+      .with(sts_auth_params)
+      .returns(assume_role_web_identity_service_instance)
+
+    assume_role_web_identity_service_instance
+      .expects(:identity_credentials)
+      .returns(sts_credentials)
+
+    assert Aws::CredentialsService.call(sts_auth_params), { credentials: sts_credentials }
+  end
+
+  test '#call returns IAM credentials when both authentication types are available' do
+    assume_role_web_identity_service_instance.expects(:identity_credentials).never
+
+    assert Aws::CredentialsService.call(full_params), iam_auth_params
+  end
+
+  test '#call raises an error if the web_identity_token_file does not exist' do
+    token_not_found_error = Aws::Sts::AssumeRoleWebIdentityService::TokenNotFoundError
+    File.stubs(:exist?).with(sts_auth_params[:web_identity_token_file]).returns(false)
+
+    assume_role_web_identity_service_instance.expects(:config).raises(token_not_found_error)
+
+    assert_raises(token_not_found_error) do
+      Aws::CredentialsService.call(sts_auth_params)
+    end
+  end
+
+  test '#call raises an error if not enough credential params are present' do
+    assert_raises(Aws::CredentialsService::AuthenticationTypeError) do
+      Aws::CredentialsService.call(full_params.except(:access_key_id, :role_arn))
+    end
+  end
+
+  private
+
+  def full_params
+    {
+      access_key_id: 'access_key_id',
+      secret_access_key: 'secret_access_key',
+      web_identity_token_file: '/path/to/token',
+      role_arn: 'role_arn',
+      role_session_name: 'role_session_name',
+      region: 'region'
+    }
+  end
+
+  def iam_auth_params
+    full_params.slice(:access_key_id, :secret_access_key)
+  end
+
+  def sts_auth_params
+    full_params.slice(:region, :role_arn, :role_session_name, :web_identity_token_file)
+  end
+
+  def sts_credentials
+    @sts_credentials ||= Aws::Credentials.new(nil, nil)
+  end
+
+  def assume_role_web_identity_service_instance
+    Aws::Sts::AssumeRoleWebIdentityService.instance
+  end
```

Comment:
`Aws::Sts::AssumeRoleWebIdentityService.instance` is used multiple times in the test, so I added this method just so there's no code duplication.

Reply:
DRY shouldn't be taken to the extreme. There are multiple sources on the internet that can be read about it. This one looks reasonable to me [1].

In this case I commented because I had to scroll down to see what this method did exactly and then I see just a single call. So it doesn't reduce complexity. Just adds some indirection. It also doesn't cache or benefit us as far as I can see.

As I wrote, just a nitpick, so not really concerned whether it stays or goes. I like theoretical discussions though :)

[1] https://betterprogramming.pub/when-dry-doesnt-work-go-wet-6befda0444bf

Reply:
I wouldn't call this extreme DRY or anything of sorts 😅
This is a Singleton so there's really no harm in calling it multiple times but that would mean, for this test, calling  `CMS::STS::AssumeRoleWebIdentity.instance` 7 times, which looks even odder to me.
This link you shared has an interesting view on how "bad abstractions" can have an impact on the future, when new criteria appears and the old abstract code needs to be changed. I don't think it applies to this piece of code on the test we're discussing, as it's just one verbose named method that returns an instance of a Singleton. If anything, this method is nothing more than a local helper.

As I mentioned, "I added this method **just** so there's no code duplication.". Now, if you think it brings more complexity than benefits, I'd be glad to remove it. Or maybe the method's name is the problem here? 🤔

Reply:
"problem" is a strong word here. Just as author you know what this method does. And I as a reader got annoyed that I have to lookup what it did only to see it is just singular call. I don't worry about repeating 7 times as it is easy to `sed` it if it ever needs to change, which is unlikely.

So just giving you the perspective of a reader. And my preference. But it is absolutely fine to keep :)

Reply:
Another thought on this: IF this was an (R)spec instead of a (Mini)test, my first idea would be to add `CMS::STS::AssumeRoleWebIdentity.instance` to a `let` variable.
Maybe this would be clearer if, instead of this private method, I'd declare an instance variable in this test's `setup` method 🤔.

Reply:
Changed it in afe6d15 :v:

Reply:
Aaaaand removed in c89d43e, as the usage of `class_attribute` in the Singleton removed the necessity of working directly with an instance when setting attributes :v: